<a href="https://colab.research.google.com/github/veniaibinder/DS-Unit-1-Sprint-2-Statistics/blob/master/module2/LS_DS_122_Chi2_Tests_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Assignment - Practice Chi^2 Tests

Use the following dataset relating to math scores of students in two different Portugese schools:

<https://archive.ics.uci.edu/ml/datasets/Student+Performance>

### 1) Load the dataset specifically relating to math scores as a new dataframe.

There are two datasets in the `student.zip` file, make sure you use `student-mat.csv`.


In [2]:
# YOUR WORK HERE
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip

--2020-06-16 23:51:34--  https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 20478 (20K) [application/x-httpd-php]
Saving to: ‘student.zip’

student.zip         100%[===================>]  20.00K  --.-KB/s    in 0.03s   

2020-06-16 23:51:34 (691 KB/s) - ‘student.zip’ saved [20478/20478]



In [3]:
!unzip student.zip

Archive:  student.zip
  inflating: student-mat.csv         
  inflating: student-por.csv         
  inflating: student-merge.R         
  inflating: student.txt             


In [4]:
from scipy import stats
import pandas as pd
import numpy as np


df = pd.read_csv('student-mat.csv', sep=';')

print(df.shape)
df.head()


(395, 33)


school sex  age address famsize Pstatus  ...  Walc  health absences  G1  G2  G3
0     GP   F   18       U     GT3       A  ...     1       3        6   5   6   6
1     GP   F   17       U     GT3       T  ...     1       3        4   5   5   6
2     GP   F   15       U     LE3       T  ...     3       3       10   7   8  10
3     GP   F   15       U     GT3       T  ...     1       5        2  15  14  15
4     GP   F   16       U     GT3       T  ...     2       5        4   6  10  10

[5 rows x 33 columns]

### 2) Use Chi^2 tests and `stats.chi2_contingency()` to identify:
 - Two pairs of variables that are dependent (are associated with one another).
 - Two pairs of variables that are independent (have no significant relationship).

Does it make intuitive sense why the variables in these pairs might or might not show a relationship to one another? 


In [ ]:
##Two pairs of variables that are dependent (are associated with one another)
df['traveltime'].value_counts().sort_index()
df['failures'].value_counts().sort_index()
contingency = pd.crosstab(df['traveltime'], df['failures'])

contingency

failures      0   1  2   3
traveltime                
1           209  30  7  11
2            80  17  9   1
3            17   3  1   2
4             6   0  0   2

In [ ]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingency)

print("chi2 statistic", chi2)
print("p value", p_value)
print("degrees of freedom",dof)
print("expected frequencies table", expected)

chi2 statistic 21.124872298979813
p value 0.012107363244231761
degrees of freedom 9
expected frequencies table [[202.99746835  32.53164557  11.06075949  10.41012658]
 [ 84.5164557   13.5443038    4.60506329   4.33417722]
 [ 18.16708861   2.91139241   0.98987342   0.93164557]
 [  6.31898734   1.01265823   0.3443038    0.32405063]]


Conclusion

Null Hypothesis: These two variables are "independent" there's no relationship between the two

Alternative Hypothesis: The two variables are "dependent" there is a relationship or association between them.

Based on the chi statistic of 21.12 and p-value of 0.012, we reject the null hypothesis that travel time and failures are independent. We suggest the alternative hypothesis.

In [ ]:
#Two pairs of variables that are independent (have no significant relationship).

df['traveltime'].value_counts().sort_index()
df['studytime'].value_counts().sort_index()
contingency = pd.crosstab(df['traveltime'], df['studytime'])

contingency

studytime    1    2   3   4
traveltime                 
1           60  133  47  17
2           31   51  16   9
3           11   11   1   0
4            3    3   1   1

In [ ]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingency)

print("chi2 statistic", chi2)
print("p value", p_value)
print("degrees of freedom",dof)
print("expected frequencies table", expected)

chi2 statistic 11.200170929144969
p value 0.26223746329595793
degrees of freedom 9
expected frequencies table [[ 68.3164557  128.82531646  42.29113924  17.56708861]
 [ 28.44303797  53.63544304  17.60759494   7.31392405]
 [  6.11392405  11.52911392   3.78481013   1.5721519 ]
 [  2.12658228   4.01012658   1.3164557    0.54683544]]


**Conclusion**

Null Hypothesis: These two variables are "independent" there's no relationship between the two

Alternative Hypothesis: The two variables are "dependent" there is a relationship or association between them.

Based on the chi statistic of 11.2 and p-value of 0.26, we fail to reject the null hypothesis that travel time and study time at home are independent. 



### 3) Use NumPy to perform your own chi^2 test "from scratch" 

Pick any of the chi^2 tests that you ran in part 2 and try them on your own. You should get the same results that Scipy got for all four values returned from `chi2_contingency()`

In [34]:
# YOUR WORK HERE
df['traveltime'].value_counts().sort_index()
df['address'].value_counts().sort_index()
observed = pd.crosstab(df['address'], df['traveltime'])

observed

traveltime    1   2   3  4
address                   
R            35  34  14  5
U           222  73   9  3

In [35]:
observed_with_margins = pd.crosstab(df['traveltime'], df['address'], margins=True)

observed_with_margins

address      R    U  All
traveltime              
1           35  222  257
2           34   73  107
3           14    9   23
4            5    3    8
All         88  307  395

In [36]:
observed = observed.values
observed

array([[ 35,  34,  14,   5],
       [222,  73,   9,   3]])

In [37]:
expected_row1_col1 = 88*257/395


expected_row1_col1


57.255696202531645

In [40]:
row_totals = df['address'].value_counts().sort_index().values

row_totals

array([ 88, 307])

In [41]:
col_totals = df['traveltime'].value_counts().sort_index().values

col_totals

array([257, 107,  23,   8])

In [23]:
sample_size = df.shape[0]

sample_size

395

In [47]:
expected = np.array([[0,0,0,0],
                     [0,0,0,0]])
for i, row_total in enumerate(row_totals): 
  for j, col_total in enumerate(col_totals):
    expected_value = (row_total*col_total / sample_size)
    # print(expected_value, i, j)
    expected[i][j] = expected_value
expected

array([[ 57,  23,   5,   1],
       [199,  83,  17,   6]])

In [48]:
observed

array([[ 35,  34,  14,   5],
       [222,  73,   9,   3]])

In [49]:
expected

array([[ 57,  23,   5,   1],
       [199,  83,  17,   6]])

In [50]:
((observed - expected)**2 / expected).sum()

55.07991425214064

## Stretch goals:

### 1. Refactor your code so it is elegant, readable, and holds reusable code in functions.

In [ ]:
# YOUR WORK HERE



### 2. Check For Understanding - Study and write your own explanations/definitions for these topics:

- What is a sample "estimate" in statistics?

- What are hypothesis test? How are they useful?

- What is a "null hypothesis?"

- What is a p-value and what does it represent?

- What does it mean for something to be "statistically significant?"

- What is a test statistic and how does it relate to a p-value?

- What are "degrees of freedom" and how are they calculated in a 1-sample, 2-sample, and chi^2 test for independence?

## Resources

- [Interactive visualize the Chi-Squared test](https://homepage.divms.uiowa.edu/~mbognar/applets/chisq.html)
- [Calculation of Chi-Squared test statistic](https://en.wikipedia.org/wiki/Pearson%27s_chi-squared_test)
- [Visualization of a confidence interval generated by R code](https://commons.wikimedia.org/wiki/File:Confidence-interval.svg)
- [Expected value of a squared standard normal](https://math.stackexchange.com/questions/264061/expected-value-calculation-for-squared-normal-distribution) (it's 1 - which is why the expected value of a Chi-Squared with $n$ degrees of freedom is $n$, as it's the sum of $n$ squared standard normals)